src/main/java/com/commercetools/sync/types/utils/TypeUpdateFieldDefinitionActionUtils.java

```diff
@@ -0,0 +1,280 @@
+    }
+
+    /**
+     * Checks if there are any field definitions which are not existing in the
+     * {@code newFieldDefinitions}. If there are, then "remove" field definition update actions are
+     * built.
+     * Otherwise, if the field definition still exists in the new field definition, then compare the field definition
+     * fields (label, etc..), and add the computed actions to the list of update actions.
+     *
+     * <p>Note: If the field type field changes, the old field definition is removed and the new field
+     *     definition is added with the new field type.
+     *
+     * @param oldFieldDefinitions       the list of old {@link FieldDefinition}s.
+     * @param newFieldDefinitions       the list of new {@link FieldDefinition}s.
+     * @return a list of field definition update actions if there are field that are not existing
+     *         in the new draft. If the field definition still exists in the new draft, then compare the field
+     *         definition fields (name, label, etc..), and add the computed actions to the list of update actions.
+     *         Otherwise, if the field definitions are identical, an empty optional is returned.
+     */
+    @Nonnull
+    private static List<UpdateAction<Type>> buildRemoveFieldDefinitionOrFieldDefinitionUpdateActions(
+            @Nonnull final List<FieldDefinition> oldFieldDefinitions,
+            @Nonnull final List<FieldDefinition> newFieldDefinitions) {
+
+        final Map<String, FieldDefinition> newFieldDefinitionsNameMap =
+                newFieldDefinitions
+                        .stream().collect(
+                        toMap(FieldDefinition::getName, fieldDefinition -> fieldDefinition,
+                                (fieldDefinitionA, fieldDefinitionB) -> {
+                                    throw new DuplicateNameException(format("Field definitions have duplicated names. "
+                                                    + "Duplicated field definition name: '%s'. "
+                                                    + "Field definitions names are expected "
+                                                    + "to be unique inside their type.",
+                                            fieldDefinitionA.getName()));
+                                }
+                        ));
+
+        return oldFieldDefinitions
+                .stream()
+                .map(oldFieldDefinition -> {
+                    final String oldFieldDefinitionName = oldFieldDefinition.getName();
+                    final FieldDefinition matchingNewFieldDefinition =
+                            newFieldDefinitionsNameMap.get(oldFieldDefinitionName);
+
+                    return ofNullable(matchingNewFieldDefinition)
+                        .map(newFieldDefinition -> {
+                            if (newFieldDefinition.getType() != null) {
+                                // field type is required so if null we let commercetools to throw exception
```

Comment:
The documentation for type is not completed I guess. trying add field definition without type it returns this 400 response -> `fieldDefinitions -> type: Missing required value`

Reply:
If it's wrong in the doc, then create a support ticket, so they can improve it.
Also if you look at the comment, it says `// field type is required so if null we let commercetools to throw exception`, but the line above, you have `if (newFieldDefinition.getType() != null) {`. So how can it ever be null if you check for non-null the line above?

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/304/commits/850246de2d1bafcfdfb033fd6e5c13fd85bd51e7

I deleted this if statement and created a documentation support ticket to CTP. Now we don't pass the error with this if statement. So its handled by the sync error callback mechanism.

``` json
io.sphere.sdk.client.ErrorResponseException: detailMessage: Request body does not contain valid JSON.
http response formatted body: {
  "statusCode" : 400,
  "message" : "Request body does not contain valid JSON.",
  "errors" : [ {
    "code" : "InvalidJsonInput",
    "message" : "Request body does not contain valid JSON.",
    "detailedErrorMessage" : "actions -> name: Missing required value"
  }, {
    "code" : "InvalidJsonInput",
    "message" : "Request body does not contain valid JSON.",
    "detailedErrorMessage" : "actions -> fieldDefinition -> type: Missing required value"
  } ]
}
```

Reply:
The same part of code is in the product type. Should it also be changed to be consistent? Also did @heshamMassoud have a look at this part? Because he clarified this part to me and he said it's good with the if condition.

Reply:
I can change it in this [PR](https://github.com/commercetools/commercetools-sync-java/issues/302) or revert the commit, it seems both of them working.. @heshamMassoud @lojzatran

Reply:
@lojzatran  According to the ticket answer for field definitions. 

> The required fields are not mentioned explicitly instead the optional fields are mentioned as optional.

Reply:
Heh weird, sometimes they have `Required` explicitly mentioned: https://docs.commercetools.com/http-api-projects-types.html#typedraft 😕 Can you tell them that in the ticket?

Reply:
The ticket already has been closed by the support. Probably, in that case, all the fields are `Required`. thus, it did not mentioned in the document.

Reply:
but we shouldn't remove the null check! Otherwise this method is quite vulnerable to NPEs.

I think we should remove this comment at all and check for null types and throw a `BuildUpdateActionException` accordingly, when the type is null.

**VIP NOTE**: the method `FieldDefinitionUpdateActionUtils#buildActions` will also throw an NPE if the type is null. It should be handled there too, as it can also be called standalone.

The notion of leaving a request to fail from CTP should only be followed on the sync process (for example, creating a new draft with a required field - we shouldn't check that field but rather leave it for the platform). However, here it is an update action build util which could also be used standalone without the sync. And if someone was to call this util method standalone (with a null type) they won't get an exception and the method would execute successfully as a false positive. (We don't even have it as a note in the Javadoc)